### PR TITLE
[JIT] script & logging for extracting IR from logs

### LINKED ISSUE #72471: Tool for capturing script-able logging output
### 🚀 The feature, motivation and pitch

This could be a useful tool for debugging & isolating performance issues with fusions.

1) in relevant places (e.g. nvfuser fusion), print logs of graphs with full shape info, so that inputs can be generated based only on the printed graph information
2) write a script that can load IR from the logs, and generate inputs to test on the IR.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72889

The script along with the GRAPH_EXPORT macro will allow for an easy way to extract IR from logs. One use case in this diff is to extract the fusion groups from nvfuser, so that the fusions can be tested individually.

Usage (e.g. for nvfuser test)

1. Write some test.py file that uses nvfuser
2. `PYTORCH_JIT_LOG_LEVEL=">>graph_fuser" python3 test.py 2>&1 | tee output.txt`
3. `python3 pytorch/scripts/jit/log_extract.py output.txt --nvfuser`

This will run with and without nvfuser to compare the output.

Alternatively, use `--output` to dump the IR so that it can be used in other applications.

Currently, only `--output` works (since generating input tensors is not supported)

Differential Revision: [D34440189](https://our.internmc.facebook.com/intern/diff/D34440189)

### CODE REVIEW

File: scripts/jit/log_extract.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import argparse
+import torch
+
+
+def extract_ir(filename: str) -> List[str]:
+    BEGIN = "<GRAPH_EXPORT>"
+    END = "</GRAPH_EXPORT>"
+    pfx = None
+    current = ""
+    graphs = []
```
Comment (eellison) [inline scripts/jit/log_extract.py:12]: nit: would this be easier as :
```
split_strs = filename.split(BEGIN)
graphs = []
for split_str in split_strs:
    end = split_str.find(END)
    assert end != -1
    graphs.append(split_str[:end])
```

File: scripts/jit/log_extract.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+
+def load_graph_and_inputs(ir: str) -> Tuple[Any, List[Any]]:
+    graph = torch._C.parse_ir(ir)
+    graph.makeMultiOutputIntoTuple()
+    inputs = []
+    for inp in graph.inputs():
+        if isinstance(inp.type(), torch._C.FloatType):
+            inputs.append(.5)
+        else:
```
Comment (eellison) [inline scripts/jit/log_extract.py:45]: let's assert this is a tensor here otherwise we'll get a confusing error message down the line... also we should just add an int input here

File: scripts/jit/log_extract.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            inputs.append(torch._C._representative_tensor(inp.type()))
+
+    func = torch._C._create_function_from_graph("forward", graph)
+    torch._C._jit_pass_erase_shape_information(func.graph)
+    return (func, inputs)
+
+
+def run_test(ir, inputs, *, warmup_runs=20, test_runs=20) -> float:
```
Comment (eellison) [inline scripts/jit/log_extract.py:53]: warmup_runs == 20 is probably a bit too much, especially if we're not running gradients. Same with 20 test runs.

File: scripts/jit/log_extract.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+def load_graph_and_inputs(ir: str) -> Tuple[Any, List[Any]]:
+    graph = torch._C.parse_ir(ir)
+    graph.makeMultiOutputIntoTuple()
+    inputs = []
+    for inp in graph.inputs():
+        if isinstance(inp.type(), torch._C.FloatType):
+            inputs.append(.5)
+        else:
+            inputs.append(torch._C._representative_tensor(inp.type()))
```
Comment (eellison) [inline scripts/jit/log_extract.py:46]: Let's assert that the tensors have requires_grad == 0 for now. If they have requires_grad == 1 then we're not currently exercising the gradient part of the graph

File: scripts/jit/log_extract.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    torch._C._jit_pass_erase_shape_information(func.graph)
+    return (func, inputs)
+
+
+def run_test(ir, inputs, *, warmup_runs=20, test_runs=20) -> float:
+    graph, _ = load_graph_and_inputs(ir)
+    for _ in range(warmup_runs):
+        graph(*inputs)
+
+    start_event = torch.cuda.Event(enable_timing=True)
```
Comment (eellison) [inline scripts/jit/log_extract.py:58]: In the future we're going to want to fork the timing based on cpu or cuda similar to https://github.com/pytorch/pytorch/blob/master/benchmarks/fuser/run_benchmarks.py#L208

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 6]
```diff
@@ -1994,6 +1995,20 @@ bool removeInplaceOperations(const std::shared_ptr<Graph>& graph) {
       graph, [&](Node* node) { return inplace_ops.count(node->kind()) != 0; });
 }
 
+void printFusionGroupsForDebug(std::shared_ptr<Graph>& graph) {
```
Comment (eellison) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:1998]: No need to re-traverse the graph, I would just add this logic here https://github.com/pytorch/pytorch/blob/a04c88f460d6dc461ae25ef7ddfc2687968d6b58/torch/csrc/jit/codegen/cuda/graph_fuser.cpp#L1543

**Review by eellison — COMMENTED** [review]
Looks great 😍  😍  😍 

Couple small comments. It's fine if this is tailored to NVFuser but in the future we're going to want to factor this out for autodiff / nnc debugging as well... fine if that doesn't happen here currently.

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34440189).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34440189).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34440189).

File: scripts/jit/log_extract.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        device=inp_type.device())
+
+
+def load_graph_and_inputs(ir: str) -> Tuple[Any, List[Any]]:
+    graph = torch._C.parse_ir(ir)
+    graph.makeMultiOutputIntoTuple()
+    inputs = []
+    for inp in graph.inputs():
+        if isinstance(inp.type(), torch._C.FloatType):
+            inputs.append(.5)
```
Comment (eellison) [inline scripts/jit/log_extract.py:55]: maybe in future diff we can do a range of values for int / float

**Review by eellison — APPROVED** [review]
😍  😍  😍

**Comment by eellison** [comment]
cc @jjsjann123 @wconstab

File: scripts/jit/log_extract.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+                continue
+            s = split_str[:end_loc]
+            pfx = split_strs[i - 1].splitlines()[-1]
+            lines = [x[len(pfx):] for x in s.splitlines(keepends=True)]
+            graphs.append(''.join(lines))
+
+    return graphs
+
+
+def make_tensor_from_type(inp_type: torch._C.TensorType):
```
Comment (eellison) [inline scripts/jit/log_extract.py:40]: Hmmm, actually this doesn't use the stride :/ so i think we need to update
